Mathlib/Computability/QueryComplexity/Defs.lean

```diff
@@ -0,0 +1,102 @@
+in this module is based on the original stochastic version. Source:
+  https://github.com/girving/debate/blob/862fdb1cf55df0d541b802bdb1e672d724df6398/Comp/Oracle.lean
+
+## Main Definitions
+
+* Comp (ι : Type) {I : Type} (s : Set I) (α : Type) : Type
+* Comp.run : Comp ι s α → (I → Oracle ι) → α × (I → ℕ)
+-/
+
+open Classical
+open Set
+noncomputable section
+
+variable {ι I : Type}
+variable {s t : Set I}
+variable {α β γ : Type}
+
+namespace QueryComplexity
+
+/-- A deterministic oracle is a map from `α` to `Bool` -/
+def Oracle (α : Type) := α → Bool
+
+/-- A deterministic computation that make decisions by querying oracles. A computation is either a
+pure value or the identifier of an oracle (`o`) drawn from a predefined set (`s`), a value to
+be queried by the oracle (`i`) and two other computations, to be run depending on the answer
+of the oracle. -/
+inductive Comp (ι : Type) {I : Type} (s : Set I) (α : Type) : Type where
+  | pure' : α → Comp ι s α
+  | query' : (o : I) → o ∈ s → (i : ι) → (Bool → Comp ι s α) → Comp ι s α
+
+namespace Comp
+
+/-- The standard bind operation for `Comp` -/
+def bind' (f : Comp ι s α) (g : α → Comp ι s β) : Comp ι s β := match f with
+  | .pure' x => g x
+  | .query' o m y f => .query' o m y fun b => (f b).bind' g
+
+/-- `Comp` is a monad -/
+instance : Monad (Comp ι s) where
+  pure := Comp.pure'
+  bind := Comp.bind'
+
+/-- Produce a `Comp` given the identifier of an oracle and a value to be queried. The `Comp`
+just returns `true` or `false` according to the answer of the oracle. -/
+def query (o : I) (y : ι) : Comp ι {o} Bool :=
+  Comp.query' o (mem_singleton _) y pure
+
+/-- Execute `f` with the oracles `os`. Returns the final value and the number of queries to
+each one of the oracles. -/
+def run (f : Comp ι s α) (os : I → Oracle ι) : α × (I → ℕ) := match f with
+  | .pure' x => (x, fun _ => 0)
+  | .query' i _ y f =>
+    let x := os i y
+    let (z,c) := (f x).run os
+    (z, c + fun j => if j = i then 1 else 0)
+
+/-- The value of a `Comp ι s` after execution -/
+def value (f : Comp ι s α) (o : I → Oracle ι) : α :=
+  (f.run o).1
+
+/-- The value of a `Comp ι s` after execution with a single oracle -/
+@[simp]
+def value' (f : Comp ι s α) (o : Oracle ι) : α :=
+  f.value fun _ ↦ o
+
+/-- The query count for a specific oracle of a `Comp ι s` -/
+def cost (f : Comp ι s α) (o : I → Oracle ι) (i : I) : ℕ :=
+  (f.run o).2 i
+
+/-- The cost of a `Comp ι s`, when run with a single oracle -/
+def cost' (f : Comp ι s α) (o : Oracle ι) : I → ℕ :=
+  f.cost fun _ ↦ o
+
+/-- Extend the set of allowed oracles in a computation -/
+def allow (f : Comp ι s α) (st : s ⊆ t) : Comp ι t α := match f with
+  | .pure' x => pure x
+  | .query' i m y f => .query' i (st m) y (fun b => (f b).allow st)
+
+/-- Extend the set of allowed oracles in a computation to the universe set -/
+def allow_all (f : Comp ι s α) : Comp ι (@univ I) α :=
+  f.allow (subset_univ s)
```

Comment:
It's not entirely clear to me what purpose `s` serves in `Comp` (which is the reason for these existing); as far as I can tell, the sorting stuff does not need it.

@girving, do you have an example in mind where this is useful, vs just changing the indexing type of `I`?

Reply:
Here’s how it’s used in the debate protocol: https://github.com/girving/debate/blob/d3e2d0bb7f6f450bb3899fb349a69133329541eb/Debate/Protocol.lean#L94

And without it, this theorem and similar wouldn’t be true, as an evil Bob could query Alice’s oracle: https://github.com/girving/debate/blob/d3e2d0bb7f6f450bb3899fb349a69133329541eb/Debate/Cost.lean#L55

I should say I’m sympathetic to the worry that this is too specific to one kind of formalized result. We don’t need it for sort, so stripping it would be fine if we just want to get sort into mathlib (my lower bound on sort doesn’t use this either, of course).

Reply:
The alternative I'm proposing in that example is for Bob, Alice, and Vera to have different indexing types. So for instance, your `def step` would have indexing type `BobUnit ⨁ AliceUnit ⨁ VeraUnit`.

Of course, this then becomes tricky to plumb combining together the oracles...

Reply:
Ah, interesting. So the set argument to
Comp goes away, and cost takes a set of queries to count across.

Reply:
My guess is that doing it with just types ends up messier, at least if you want to do variety of subsets. But I’m not confident about this.

Reply:
I'm happy to stick with the sets for now.

Reply:
I had a play around with a set-free version in #22194